examples/step-63/step-63.cc

```diff
@@ -19,16 +19,1091 @@
+                         const unsigned int component = 0) const;
+
+    virtual void value_list(const std::vector<Point<dim>> &points,
+                            std::vector<double> &          values,
+                            const unsigned int             component = 0) const;
+  };
+
+
+  template <int dim>
+  double BoundaryValues<dim>::value(const Point<dim> & p,
+                                    const unsigned int component) const
+  {
+    Assert(component == 0, ExcIndexRange(component, 0, 1));
+    (void)component;
+
+    if (std::fabs(p[0] * p[0] + p[1] * p[1] - 0.3 * 0.3) <
+          1e-8                                        // around cylinder
+        || std::fabs(p[0] + 1) < 1e-8                 // x == -1
+        || std::fabs(p[1] - 1) < 1e-8                 // y == 1
+        || (std::fabs(p[1] + 1) < 1e-8 && p[0] < 0.5) // y == -1, x <= 0.5
+    )
+      {
+        return 0.0;
+      }
+    else if (std::fabs(p[0] - 1) < 1e-8                     // x = 1
+             || (std::fabs(p[1] + 1) < 1e-8 && p[0] >= 0.5) // y == -1, x > 0.5
+    )
+      {
+        return 1.0;
+      }
+    else
+      {
+        return 0.0;
+      }
+  }
+
+
+
+  template <int dim>
+  void BoundaryValues<dim>::value_list(const std::vector<Point<dim>> &points,
+                                       std::vector<double> &          values,
+                                       const unsigned int component) const
+  {
+    Assert(values.size() == points.size(),
+           ExcDimensionMismatch(values.size(), points.size()));
+
+    for (unsigned int i = 0; i < points.size(); ++i)
+      values[i] = BoundaryValues<dim>::value(points[i], component);
+  }
+
+  template <int dim>
+  double delta_value(const double         hk,
+                     const double         eps,
+                     const Tensor<1, dim> dir,
+                     const double         pk)
+  {
+    // Value defined in 'On discontinuity–capturing methods for
+    // convection–diffusion equations'
+    double Peclet = dir.norm() * hk / (2.0 * eps * pk);
+    double coth =
+      (1.0 + std::exp(-2.0 * Peclet)) / (1.0 - std::exp(-2.0 * Peclet));
+
+    return hk / (2.0 * dir.norm() * pk) * (coth - 1.0 / Peclet);
+  }
+
+
+  template <int dim>
+  class AdvectionProblem
+  {
+  public:
+    AdvectionProblem(Settings settings);
+    void run();
+
+  private:
+    void setup_system();
+
+    template <class IteratorType>
+    void assemble_cell(const IteratorType &cell,
+                       ScratchData<dim> &  scratch_data,
+                       CopyData &          copy_data);
+    void assemble_system_and_multigrid();
+
+    std::unique_ptr<MGSmoother<Vector<double>>> create_smoother();
+
+    void solve();
+    void refine_grid();
+    void output_results(const unsigned int cycle) const;
+
+    Triangulation<dim> triangulation;
+    DoFHandler<dim>    dof_handler;
+
+    FE_Q<dim>     fe;
+    MappingQ<dim> mapping;
+    unsigned int  quad_degree;
```

Comment:
let's spell these things out: `quadrature_degree`. Although we pretty universally just use `fe.degree+1`. Is there a reason to have a separate variable for it here?

Reply:
I suspect that you can mark some of these variables as `const` if they are initialized in the constructor.

Reply:
No, I only use `quadrature_degree` twice in the code, so I will remove the separate variable